source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -554,6 +614,400 @@ namespace aspect
+        // This if statement is only there for the unit test. In normal sitations it should always be set to false,
+        // because the nondimensionalization should always be done (in this exact way), unless you really know what
+        // you are doing.
+        double nondimensionalization_value = 1.0;
+        if (!prevent_nondimensionalization)
+          {
+            const std::array< double, 3 > eigenvalues = dealii::eigenvalues(strain_rate_3d);
+            nondimensionalization_value = std::max(std::abs(eigenvalues[0]),std::abs(eigenvalues[2]));
+
+            Assert(!std::isnan(nondimensionalization_value), ExcMessage("The second invariant of the strain rate is not a number."));
+
+            // Make the strain-rate and velocity gradient tensor non-dimensional
+            // by dividing it through the second invariant
+            if (nondimensionalization_value != 0)
+              {
+                strain_rate_nondimensional /= nondimensionalization_value;
+                velocity_gradient_tensor_nondimensional /=  nondimensionalization_value;
+              }
+          }
+
+        // create output variables
+        std::vector<double> deriv_volume_fractions(n_grains);
+        std::vector<Tensor<2,3>> deriv_a_cosine_matrices(n_grains);
+
+        // create shorcuts
+        const std::array<double, 4> &tau = ref_resolved_shear_stress;
+
+        std::vector<double> strain_energy(n_grains);
+        double mean_strain_energy = 0;
+
+        // loop over grains
+        for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+          {
+            // Compute the Schmidt tensor for this grain (nu), s is the slip system.
+            // We first compute beta_s,nu (equation 5, Kaminski & Ribe, 2001)
+            // Then we use the beta to calculate the Schmidt tensor G_{ij} (Eq. 5, Kaminski & Ribe, 2001)
+            Tensor<2,3> G;
+            Tensor<1,3> w;
+            Tensor<1,4> beta({1.0, 1.0, 1.0, 1.0});
+
+            // these are variables we only need for olivine, but we need them for both
+            // within this if bock and the next ones
+            // todo: initialize to dealii uninitialized value
+            unsigned int index_max_q = 0;
+            unsigned int index_intermediate_q = 0;
+            unsigned int index_min_q = 0;
+            unsigned int index_inactive_q = 0;
+
+            // compute G and beta
+            Tensor<1,4> bigI;
+            // this should be equal to a_cosine_matrices[grain_i]*a_cosine_matrices[grain_i]?
+            // todo: check and maybe replace?
+            const Tensor<2,3> rotation_matrix = get_rotation_matrix_grains(cpo_index,data,mineral_i,grain_i);
+            for (unsigned int i = 0; i < 3; ++i)
+              {
+                for (unsigned int j = 0; j < 3; ++j)
+                  {
+                    bigI[0] = bigI[0] + strain_rate_nondimensional[i][j] * rotation_matrix[0][i] * rotation_matrix[1][j];
```

Comment:
Is there maybe a tensor operator that does this multiplication?

Reply:
Yes, it is taking the outer product of the slip system normal and slip direction, and then taking the scalar product of that with the non dimensional strain-rate. I have replaced the code and the test still works :)